azure/durable_functions/models/TaskOrchestrationExecutor.py

```diff
@@ -0,0 +1,280 @@
+        # this mapping is used for processing events that transition a Task from its running state
+        # to a terminal one
+        SetTaskValuePayload = namedtuple("SetTaskValuePayload", ("is_success", "task_id_key"))
+        self.event_to_SetTaskValuePayload = dict([
+            (HistoryEventType.TASK_COMPLETED, SetTaskValuePayload(True, "TaskScheduledId")),
+            (HistoryEventType.TIMER_FIRED, SetTaskValuePayload(True, "TimerId")),
+            (HistoryEventType.SUB_ORCHESTRATION_INSTANCE_COMPLETED,
+             SetTaskValuePayload(True, "TaskScheduledId")),
+            (HistoryEventType.EVENT_RAISED, SetTaskValuePayload(True, "Name")),
+            (HistoryEventType.TASK_FAILED, SetTaskValuePayload(False, "TaskScheduledId")),
+            (HistoryEventType.SUB_ORCHESTRATION_INSTANCE_FAILED,
+             SetTaskValuePayload(False, "TaskScheduledId"))
+        ])
+        self.task_completion_events = set(self.event_to_SetTaskValuePayload.keys())
+        self.initialize()
+
+    def initialize(self):
+        """Initialize the TaskOrchestrationExecutor for a new orchestration invocation."""
+        # The first task is just a placeholder to kickstart the generator.
+        # So it's value is `None`.
+        # TODO: need to set the `is_replaying` flag in here!
+        self.current_task: TaskBase = AtomicTask(-1, [])
+        self.current_task.set_value(is_error=False, value=None)
+
+        self.output: Any = None
+        self.exception: Optional[Exception] = None
+        self.orchestrator_returned: bool = False
+
+    def execute(self, context: DurableOrchestrationContext,
+                history: List[HistoryEvent], fn) -> str:
+        """Execute an orchestration using the orchestration history to evaluate Tasks and replay events.
+
+        Parameters
+        ----------
+        context : DurableOrchestrationContext
+            The user's orchestration context, to interact with the user code.
+        history : List[HistoryEvent]
+            The orchestration history, to evaluate tasks and replay events.
+        fn : function
+            The user's orchestration function.
+
+        Returns
+        -------
+        str
+            A JSON-formatted string of the user's orchestration state, payload for the extension.
+        """
+        self.context = context
+        evaluated_user_code = fn(context)
+
+        # If user code is a generator, then it uses `yield` statements (the DF API)
+        # and so we iterate through the DF history, generating tasks and populating
+        # them with values when the history provides them
+        if isinstance(evaluated_user_code, GeneratorType):
+            self.generator = evaluated_user_code
+            for event in history:
+                self.process_event(event)
+                if self.has_execution_completed:
+                    break
+
+        # Due to backwards compatibility reasons, it's possible
```

Comment:
Ah yes, you're right. In the current release, we support both, in part because, in a early releases of the Python and JavaScript SDKs, we required customers to use `yield` even in the `continue_as_new` case.

If we were to somehow decide to make the release containing these changes into a major release, I could definitely add some checks to force users to use `continue_as_new` without `yield`. If the performance gains are dramatic enough, I think there would be a good enough justification to do this.

Reply:
I think I'd rather tie major releases more to a change in programming model (from generators to promises or something like that), but I'm not 100% against the idea (especially for JavaScript where we have had many 1.x releases already).